File: Client_Root/Client/Assets/Scripts/Network/Messages/ToRoom/GameInputSkillToR.cs
using UnityEngine;
using FlatBuffers;
using System.Collections.Generic;

public class GameInputSkillToR : IMessage
{
    public const ushort MESSAGE_ID = MessageID.GameInputSkillToR_ID;

    public int m_nPlayerIndex;
    public int m_nSkillID;
    public List<Vector3> m_listVector3 = new List<Vector3>();
    public List<int> m_listInt = new List<int>();
    public List<float> m_listFloat = new List<float>();
    public FBS.InputType m_InputType;

    public override ushort GetID()
    {
        return MESSAGE_ID;
    }

    public override IMessage Clone()
    {
        return null; 
    }

    public override byte[] Serialize()
    {
        FBS.GameInputSkillToR.StartVector3sVector(m_Builder, m_listVector3.Count);
        foreach(Vector3 vec3 in m_listVector3)
        {
            FBS.Data.Vector3.CreateVector3(m_Builder, vec3.x, vec3.y, vec3.z);
        }
        var vector3s = m_Builder.EndVector();
        var ints = FBS.GameInputSkillToR.CreateIntsVector(m_Builder, m_listInt.ToArray());
        var floats = FBS.GameInputSkillToR.CreateFloatsVector(m_Builder, m_listFloat.ToArray());

        FBS.GameInputSkillToR.StartGameInputSkillToR(m_Builder);
        FBS.GameInputSkillToR.AddPlayerIndex(m_Builder, m_nPlayerIndex);
        FBS.GameInputSkillToR.AddSkillID(m_Builder, m_nSkillID);
        FBS.GameInputSkillToR.AddVector3s(m_Builder, vector3s);
        FBS.GameInputSkillToR.AddInts(m_Builder, ints);
        FBS.GameInputSkillToR.AddFloats(m_Builder, floats);
        FBS.GameInputSkillToR.AddInput(m_Builder, m_InputType);
        var data = FBS.GameInputSkillToR.EndGameInputSkillToR(m_Builder);

        m_Builder.Finish(data.Value);

        return m_Builder.SizedByteArray();
    }

    public override bool Deserialize(byte[] bytes)
    {
        var buf = new ByteBuffer(bytes);

        var data = FBS.GameInputSkillToR.GetRootAsGameInputSkillToR(buf);

        m_nPlayerIndex = data.PlayerIndex;
        m_nSkillID = data.SkillID;
        for (int i = 0; i < data.Vector3sLength; ++i)
        {
            FBS.Data.Vector3 vec3 = data.GetVector3s(i);
            m_listVector3.Add(new Vector3(vec3.X, vec3.Y, vec3.Z));
        }
        for (int i = 0; i < data.IntsLength; ++i)
        {
            m_listInt.Add(data.GetInts(i));
        }
        for (int i = 0; i < data.FloatsLength; ++i)
        {
            m_listFloat.Add(data.GetFloats(i));
        }
        m_InputType = data.Input;

        return true;
    }

    public override void OnReturned()
    {
        m_listVector3.Clear();
        m_listInt.Clear();
        m_listFloat.Clear();
    }
}

File: Server_Root/Server/RoomServer/Messages/ToRoom/GameInputSkillToR.cpp
#include "stdafx.h"
#include "GameInputSkillToR.h"

GameInputSkillToR::GameInputSkillToR()
{
}

GameInputSkillToR::~GameInputSkillToR()
{
}

unsigned short GameInputSkillToR::GetID()
{
	return MESSAGE_ID;
}

IMessage* GameInputSkillToR::Clone()
{
	GameInputSkillToR* pClone = new GameInputSkillToR();

	pClone->m_nPlayerIndex = m_nPlayerIndex;
	pClone->m_nSkillID = m_nSkillID;
	pClone->m_vecVector3 = m_vecVector3;
	pClone->m_vecInt = m_vecInt;
	pClone->m_vecFloat = m_vecFloat;
	pClone->m_InputType = m_InputType;

	return pClone;
}

const char* GameInputSkillToR::Serialize(int* pLength)
{
	vector<FBS::Data::Vector3> vecVector3;
	for (vector<btVector3>::iterator it = m_vecVector3.begin(); it != m_vecVector3.end(); ++it)
	{
		btVector3 vec3 = *it;
		vecVector3.push_back(FBS::Data::Vector3(vec3.x(), vec3.y(), vec3.z()));
	}

	auto vector3s = m_Builder.CreateVectorOfStructs(vecVector3);
	auto ints = m_Builder.CreateVector(m_vecInt);
	auto floats = m_Builder.CreateVector(m_vecFloat);

	FBS::GameInputSkillToRBuilder data_builder(m_Builder);
	data_builder.add_PlayerIndex(m_nPlayerIndex);
	data_builder.add_SkillID(m_nSkillID);
	data_builder.add_Vector3s(vector3s);
	data_builder.add_Ints(ints);
	data_builder.add_Floats(floats);
	data_builder.add_Input(m_InputType);
	auto data = data_builder.Finish();

	m_Builder.Finish(data);

	*pLength = m_Builder.GetSize();

	return (char*)m_Builder.GetBufferPointer();
}

bool GameInputSkillToR::Deserialize(const char* pChar)
{
	auto data = flatbuffers::GetRoot<FBS::GameInputSkillToR>((const void*)pChar);

	m_nPlayerIndex = data->PlayerIndex();
	m_nSkillID = data->SkillID();
	for (int i = 0; i < data->Vector3s()->size(); ++i)
	{
		const FBS::Data::Vector3* vec3 = data->Vector3s()->Get(i);
		m_vecVector3.push_back(btVector3(vec3->x(), vec3->y(), vec3->z()));
	}
	for (int i = 0; i < data->Ints()->size(); ++i)
	{
		m_vecInt.push_back(data->Ints()->Get(i));
	}
	for (int i = 0; i < data->Floats()->size(); ++i)
	{
		m_vecFloat.push_back(data->Floats()->Get(i));
	}
	m_InputType = data->Input();

	return true;
}